### PR TITLE
[nightly-test] add test shuffle_data_loader

### PR DESCRIPTION
**Pull request by scv119**
add nightly test for running @clarkzinzow shuffle data loader benchmark.

### CODE REVIEW

**Comment by rkooo567** [comment]
btw, why is this do not merge?

**Comment by scv119** [comment]
@rkooo567  s3 permission issues...

File: release/nightly_tests/nightly_tests.yaml [hunk 1]
```diff
@@ -242,3 +242,13 @@
...
+
+# Stress test shuffle_data_loader.
+- name: shuffle_data_loader
+  cluster:
+    app_config: shuffle_data_loader/shuffle_data_loader_app_config.yaml
+    compute_template: shuffle_data_loader/shuffle_data_loader_compute.yaml
+
+  run:
+    timeout: 7200 
+    script: bash shuffle_data_loader/shuffle_data_loader_test.sh
```
Comment (rkooo567) [inline release/nightly_tests/nightly_tests.yaml:254]: Newline

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_app_config.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+post_build_cmds:
+  - sudo apt-get install -y unzip
+  - cd ~ && wget https://codeload.github.com/ray-project/ray_shuffling_data_loader/zip/refs/heads/feat/benchmarks
+  - cd ~ && unzip benchmarks
+  - pip uninstall -y ray
+  - pip3 install -U {{ env["RAY_WHEELS"] | default("ray") }}
+  - pip3 install -U ray[default]
+  - pip install -U git+https://github.com/ray-project/ray_shuffling_data_loader.git@main#egg=ray_shuffling_data_loader
+  - pip install -U --user numpy pandas pyarrow fastparquet s3fs
```
Comment (rkooo567) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_app_config.yaml:17]: Newline here!

Comment (rkooo567) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_app_config.yaml:17]: This can be added to the `pip_packages` section

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+head_node_type:
+    name: head_node
+    instance_type: i3.4xlarge
+    resources: {"object_store_memory": 53687091200}
+
+worker_node_types:
+    - name: worker_node
+      instance_type: i3.4xlarge
+      min_workers: 0
```
Comment (rkooo567) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml:20]: @clarkzinzow is this supposed to test a single node? I thought we should test multi node.

**Review by rkooo567 — COMMENTED** [review]
@clarkzinzow can you tell us the configuration to invoke object spilling & multi node? Maybe we can have 2 tests (single node lightweight + more complex case )

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_test.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                                        --batch-size "$batch_size" \
+                                        --num-trials "$num_trials" \
+                                        --cluster \
+                                        --num-reducers "$num_reducers" \
+                                        --num-trainers "$num_trainers" \
+                                        --num-epochs "$num_epochs" \
+                                        --max-concurrent-epochs "$max_concurrent_epochs" \
+                                        --data-dir "$data_dir" \
+                                        --stats-dir "$stats_dir" \
+                                        --unique-stats
```
Comment (clarkzinzow) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_test.sh:38]: Assuming that we don't want to save stats to S3 for every run, I think that we should remove `--stats-dir` and `--unique-stats`, and add `--no-stats`.

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_app_config.yaml (hunk 2 above)
Comment (clarkzinzow) [reply release/nightly_tests/shuffle_data_loader/shuffle_data_loader_app_config.yaml:17]: When adding to the `pip_packages` section, I'd recommend not including `fastparquet`. I think we'll want to use `pyarrow` instead, and `pyarrow` will only be used if `fastparquet` is not installed.

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml (hunk 3 above)
Comment (clarkzinzow) [reply release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml:20]: I think that we should have a simple single-node test (smaller data scale e.g. 100M rows) and a large-scale multi-node test (4 nodes total, 4 trainers per node, 32 reducers total, 400M rows, 50 files).

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_test.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+num_files_list=(25)
+num_trainers_list=(4)
+num_reducers_per_trainer_list=(2)
+
+for max_concurrent_epochs in "${max_concurrent_epochs_list[@]}"; do
+        for num_files in "${num_files_list[@]}"; do
+                for num_trainers in "${num_trainers_list[@]}"; do
+                        for num_reducers_per_trainer in "${num_reducers_per_trainer_list[@]}"; do
+                                num_reducers=$(( num_reducers_per_trainer * num_trainers ))
+                                python ~/ray_shuffling_data_loader-feat-benchmarks/benchmarks/benchmark.py \
```
Comment (clarkzinzow) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_test.sh:25]: This will need to be changed to `ray submit /path/to/cluster/config.yaml /path/to/benchmark.py` or `ray exec /path/to/cluster/config.yaml "python /path/to/mounted/dir/benchmark.py"` when running on a multi-node cluster.

**Comment by scv119** [comment]
have problem moving the s3 bucket. might need @thomasdesr 's help.

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml (hunk 3 above)
Comment (clarkzinzow) [reply release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml:20]: But if we had to choose, the large-scale multi-node test is more important.

**Comment by scv119** [comment]
https://beta.anyscale.com/o/anyscale-internal/projects/prj_SVFGM5yBqK6DHCfLtRMryXHM/clusters/ses_87SxsphLiyCJnTKhtrEwcXWb seems to be successful?

File: release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+head_node_type:
+    name: head_node
+    instance_type: i3.4xlarge
+    resources: {"object_store_memory": 53687091200}
+
+worker_node_types:
+    - name: worker_node
+      instance_type: m5.4xlarge
+      min_workers: 0
+      max_workers: 4
```
Comment (rkooo567) [inline release/nightly_tests/shuffle_data_loader/shuffle_data_loader_compute.yaml:21]: Revert this to 0 if this is for a single node?

**Review by rkooo567 — APPROVED** [review]
One last comment

**Comment by rkooo567** [comment]
This just changes the test config. Test failures should be unrelated.
